src/metabase/util/queue.clj

```diff
@@ -63,24 +70,155 @@
+  [listener-name :- :string
+   queue :- (ms/InstanceOfClass BlockingQueue)
+   handler :- [:=> [:cat [:sequential :any]] :any]
+   {:keys [success-handler
+           err-handler
+           pool-size
+           max-batch-messages
+           max-next-ms]
+    :or   {success-handler (constantly nil)
+           err-handler (constantly nil)
+           pool-size       1
+           max-batch-messages 50
+           max-next-ms     100}} :- ::listener-options]
+  (if (listener-exists? listener-name)
+    (log/errorf "Listener %s already exists" listener-name)
+
+    (let [executor (Executors/newFixedThreadPool
+                    pool-size
+                    (.build
+                     (doto (BasicThreadFactory$Builder.)
+                       (.namingPattern (str "queue-" listener-name "-%d"))
+                       (.daemon true))))]
+      (log/infof "Starting listener %s with %d threads %s" (u/format-color 'green listener-name) pool-size (u/emoji "\uD83C\uDFA7"))
+      (dotimes [_ pool-size]
+        (.submit ^ExecutorService executor ^Callable #(listener-thread listener-name queue handler
+                                                                       {:success-handler    success-handler
+                                                                        :err-handler        err-handler
+                                                                        :max-batch-messages max-batch-messages
+                                                                        :max-next-ms        max-next-ms})))
+
+      (swap! listeners assoc listener-name executor))))
+
+(mu/defn stop-listening!
+  "Stops the listener previously started with (listen!).
+  If there is no running listener with the given name, it is a no-op"
+  [listener-name :- :string]
+  (if-let [executor (get @listeners listener-name)]
+    (do
+      (log/infof "Stopping listener %s..." listener-name)
+      (.shutdownNow ^ExecutorService executor)
+      (.awaitTermination ^ExecutorService executor 30 TimeUnit/SECONDS)
+
+      (swap! listeners dissoc listener-name)
+      (log/infof "Stopping listener %s...done" listener-name))
+    (log/infof "No running listener named %s" listener-name)))
+
+(defmulti init-listener!
+  "Initialize a listener with a given name. All implementations of this method are called once and only
+  once when the server is starting up. Task namespaces (`metabase.*.task`) should add new
+  implementations of this method to start new listeners they define (i.e., with a call to `queue/listen!`.)
+
+  The dispatch value for this function can be any unique keyword, but by convention is a namespaced keyword version of
+  the name of the listener being initialized; for sake of consistency with the listener name itself, the keyword should be left
+  CamelCased.
+
+    (defmethod qeueue/listener-init! ::ExampleListener [_]
+      (queue/listen! (queue/listen! \"example-listener\" queue handler))"
```

Comment:
This example in the docstring is very confusing. It's calling `queue/listen!` on the return value of `queue/listen!` but with a single arity.